packages/beacon-node/src/chain/opPools/aggregatedAttestationPool.ts

```diff
@@ -547,26 +627,74 @@ export class MatchingDataAttestationGroup {
 
     this.attestations.push(attestation);
 
+    const maxRetained = isForkPostElectra(this.config.getForkName(this.data.slot))
+      ? MAX_RETAINED_ATTESTATIONS_PER_GROUP_ELECTRA
+      : MAX_RETAINED_ATTESTATIONS_PER_GROUP;
+
     // Remove the attestations with less participation
-    if (this.attestations.length > MAX_RETAINED_ATTESTATIONS_PER_GROUP) {
+    if (this.attestations.length > maxRetained) {
+      // ideally we should sort by effective balance but there is no state/effectiveBalance here
+      // it's rare to see > 8 attestations per group in electra anyway
       this.attestations.sort((a, b) => b.trueBitsCount - a.trueBitsCount);
-      this.attestations.splice(
-        MAX_RETAINED_ATTESTATIONS_PER_GROUP,
-        this.attestations.length - MAX_RETAINED_ATTESTATIONS_PER_GROUP
-      );
+      this.attestations.splice(maxRetained, this.attestations.length - maxRetained);
     }
 
     return InsertOutcome.NewData;
   }
 
   /**
    * Get AttestationNonParticipant for this groups of same attestation data.
-   * @param notSeenAttestingIndices not seen attestting indices, i.e. indices in the same committee
+   * @param notSeenCommitteeMembers not seen attestting indices, i.e. indices in the same committee
    * @returns an array of AttestationNonParticipant
    */
-  getAttestationsForBlock(fork: ForkName, notSeenAttestingIndices: Set<number>): AttestationNonParticipant[] {
+  getAttestationsForBlock(
+    fork: ForkName,
+    effectiveBalanceIncrements: EffectiveBalanceIncrements,
+    notSeenAttestingIndices: Set<number>,
+    maxAttestation: number
+  ): AttestationNonParticipant[] {
     const attestations: AttestationNonParticipant[] = [];
+    const excluded = new Set<Attestation>();
+    for (let i = 0; i < maxAttestation; i++) {
+      const mostValuableAttestation = this.getMostValuableAttestation(
+        fork,
+        effectiveBalanceIncrements,
+        notSeenAttestingIndices,
+        excluded
+      );
+
+      if (mostValuableAttestation === null) {
+        // stop looking for attestation because all attesters are seen or no attestation has missing attesters
+        break;
+      }
+
+      attestations.push(mostValuableAttestation);
+      excluded.add(mostValuableAttestation.attestation);
+      // this will narrow down the notSeenAttestingIndices for the next iteration
+      notSeenAttestingIndices = mostValuableAttestation.notSeenAttendingIndices;
```

Comment:
just to clarify from PR description
> This means there is no useless attestations included

this is because we update seen attesting indices here and then re-evaluate again, right?

Reply:
yes exactly 👍 
the next iteration only scan through `notSeenAttestingIndices` so usually it will not take much time
however it will take time for unfinality network where there is too few participation, but in that case we preaggregate aggregated attestations and bound the total attestations per group anyway